### PR TITLE
cake 5: Add native type hints for Validation

### PR DESCRIPTION
**Pull request by othercorey**

### CODE REVIEW

File: src/Validation/Validation.php [hunk 1]
```diff
@@ -324,14 +324,14 @@ public static function numElements($check, string $operator, int $expectedCount)
...
+     * @param mixed $check1 The left value to compare.
      * @param string $operator Can be one of following operator strings:
      *   '>', '<', '>=', '<=', '==', '!=', '===' and '!=='. You can use one of
      *   the Validation::COMPARE_* constants.
-     * @param string|int $check2 The right value to compare.
+     * @param mixed $check2 The right value to compare.
      * @return bool Success
      */
-    public static function comparison($check1, string $operator, $check2): bool
+    public static function comparison(mixed $check1, string $operator, mixed $check2): bool
```
Comment (othercorey) [inline src/Validation/Validation.php:334]: This is due to compareFields().

File: src/Validation/Validation.php [hunk 2]
```diff
@@ -677,7 +677,7 @@ public static function localizedTime($check, string $type = 'datetime', $format
      * @param array $booleanValues List of valid boolean values, defaults to `[true, false, 0, 1, '0', '1']`.
      * @return bool Success.
      */
-    public static function boolean($check, array $booleanValues = []): bool
+    public static function boolean(string|int|bool $check, array $booleanValues = []): bool
```
Comment (markstory) [inline src/Validation/Validation.php:680]: This should probably be mixed as well. We could get anything from end user request data.

File: src/Validation/Validation.php [hunk 3]
```diff
@@ -695,7 +695,7 @@ public static function boolean($check, array $booleanValues = []): bool
      * @param array $truthyValues List of valid truthy values, defaults to `[true, 1, '1']`.
      * @return bool Success.
      */
-    public static function truthy($check, array $truthyValues = []): bool
+    public static function truthy(string|int|bool $check, array $truthyValues = []): bool
```
Comment (markstory) [inline src/Validation/Validation.php:698]: Same here with using `mixed`.

File: src/Validation/Validation.php [hunk 4]
```diff
@@ -713,7 +713,7 @@ public static function truthy($check, array $truthyValues = []): bool
      * @param array $falseyValues List of valid falsey values, defaults to `[false, 0, '0']`.
      * @return bool Success.
      */
-    public static function falsey($check, array $falseyValues = []): bool
+    public static function falsey(string|int|bool $check, array $falseyValues = []): bool
```
Comment (markstory) [inline src/Validation/Validation.php:716]: Same here with using `mixed`

File: src/Validation/Validation.php [hunk 5]
```diff
@@ -1205,7 +1205,7 @@ public static function luhn($check): bool
      * @throws \RuntimeException when mime type can not be determined.
      * @throws \LogicException when ext/fileinfo is missing
      */
-    public static function mimeType($check, $mimeTypes = []): bool
+    public static function mimeType(UploadedFileInterface|array|string $check, array|string $mimeTypes = []): bool
```
Comment (markstory) [inline src/Validation/Validation.php:1208]: This should probably be mixed as well. Well formed requests should only have `UploadedFileInterface|array|string` but bad ones could have arbitrary garbage in them.

File: src/Validation/Validation.php [hunk 6]
```diff
@@ -839,7 +827,7 @@ public static function equalTo($check, $comparedTo): bool
      * @param array<string> $extensions file extensions to allow. By default extensions are 'gif', 'jpeg', 'png', 'jpg'
      * @return bool Success
      */
-    public static function extension($check, array $extensions = ['gif', 'jpeg', 'png', 'jpg']): bool
+    public static function extension(UploadedFileInterface|array|string $check, array $extensions = ['gif', 'jpeg', 'png', 'jpg']): bool
```
Comment (othercorey) [inline src/Validation/Validation.php:830]: Should functions like extension accept mixed also?

Comment (ADmad) [reply src/Validation/Validation.php]: Validation function take user input so it could be of any type.

Comment (othercorey) [reply src/Validation/Validation.php]: This is true. However, I've had to adjust only a few tests due to restricting types. I wonder if we're properly testing "supported" input in validation outside of a few specific exceptions.

Comment (othercorey) [reply src/Validation/Validation.php]: For example, this code certainly expects the type to be at least string and does not early return otherwise.

Comment (markstory) [reply src/Validation/Validation.php]: I think so. Validation could be passed anything as it gets end user data.

Comment (othercorey) [reply src/Validation/Validation.php]: Updated extension() to mixed and  few others. Added a note in the docblock about supported values.

**Comment by othercorey** [comment]
@markstory @ADmad last call. There are a lot of methods so I left the PR open for longer.

**Pull request merged**